### DISCUSSION TITLE
Adjoint solver performance on HPC dependent on number of processors

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#379] [General]
Hi,

I have managed to build DAFoam on my University HPC cluster. Find below the modules that I load:
```
  1) git/2.24.1               3) libfabric/1.7.0   5) boost/1.70.0   7) python/3.7.3           9) intel-mkl/2018.1.163
  2) intel-compilers/18.0.3   4) gcc/6.4.0         6) petsc/3.18.2   8) intel-mpi/2018.3.222
```
DAFoam version: 3.0.3
IPOPT version: 3.13.2

I modified the field inversion case slightly and ran the initial set of simulations on my workstation without any issue.

But when I do the same changes and run it on HPC, I am faced with two problems
1. I am not able to run on 8 cores. I am able to run on 4,16, 32 but just not 8. Not sure why.
2. My adjoint solver does not converge when I run the test case on more than 4 cores. I checked the previous threads on this problem and tried implementing the changes but nothing worked. Initially I changed ```"pcFillLevel"``` from 2 to 1 and it was working. Next I made these changes from 
```
"mu_strategy": "adaptive",

```
to

```
"mu_strategy": "adaptive",
 "mu_oracle": "probing",
 "fixed_mu_oracle": "probing",
```
and it seems to be working. 

I am trying to understand - why are points 1 and 2 happening? Does the number of cores influence the solution of the adjoint linear system?

**Comment by friedenhe** [comment]
Please upload the log files.

**Reply by uttamcadambi07** [reply]
Hello Dr. He

I managed to fix the problem of the adjoint solver. Turns out, it was in the settings of IPOPT. 

But I am attaching the log files for two cases - one with 4 cores and another with 16 cores.

My primal solution is failing in the 16 cores case (but this is not happening in the 4 cores case). All settings are exactly the same. 

Find attached the zipped folder containing the log files.
1. 4 cores - log_file_4C
2. 16 cores - log_file_16C

[log_file_cases.zip](https://github.com/mdolab/dafoam/files/10268300/log_file_cases.zip)

Search for the string "failed!" in both files and you will see that it does not exist in the log_file_4C file but does exist (around 121 times) in the log_file_16C file. The primal solver is failing to converge for the same prescribed tolerance and same maximum number of iterations when run on 16 cores (but converges on 4 cores).

**Reply by friedenhe** [reply]
The periodic hill case is known to have issues when running in parallel (because it has periodic boundaries; we haven't gotten a chance to debug it yet). Try to run this case in serial.  Note that the parallel run issues happen only for the periodic hill case, all other cases (e.g., S809) can run in parallel without an issue.

**Reply by uttamcadambi07** [reply]
Thanks Dr. He. I will try serial case and keep this thread updated.
